File: php/src/detectors/SqlInjectionDetector.php
<?php

namespace FireWTWall\Detectors;

class SqlInjectionDetector
{
    private static array $rules = [
        // Critical
        ['name' => 'sql-union-select',      'severity' => 'critical', 'pattern' => '/\bunion\s+(?:all\s+)?select\b/i'],
        ['name' => 'sql-drop-table',        'severity' => 'critical', 'pattern' => '/;\s*drop\s+table\b/i'],
        ['name' => 'sql-xp-cmdshell',       'severity' => 'critical', 'pattern' => '/\bxp_cmdshell\b/i'],
        ['name' => 'sql-exec',              'severity' => 'critical', 'pattern' => '/\bexec(?:ute)?\s*\(/i'],
        ['name' => 'sql-information-schema','severity' => 'critical', 'pattern' => '/\binformation_schema\b/i'],
        ['name' => 'sql-sleep',             'severity' => 'critical', 'pattern' => '/\bsleep\s*\(\s*\d/i'],
        ['name' => 'sql-benchmark',         'severity' => 'critical', 'pattern' => '/\bbenchmark\s*\(/i'],
        ['name' => 'sql-load-file',         'severity' => 'critical', 'pattern' => '/\bload_file\s*\(/i'],
        ['name' => 'sql-into-outfile',      'severity' => 'critical', 'pattern' => '/\binto\s+(?:out|dump)file\b/i'],
        ['name' => 'sql-sys-tables',        'severity' => 'critical', 'pattern' => '/\bsysobjects\b|\bsyscolumns\b/i'],
        // High
        ['name' => 'sql-comment',           'severity' => 'high',     'pattern' => '/(?:--|\/\*|\*\/|#\s*$)/m'],
        ['name' => 'sql-stacked-query',     'severity' => 'high',     'pattern' => '/;\s*(?:select|insert|update|delete|drop|alter|create|exec)\b/i'],
        ['name' => 'sql-cast-convert',      'severity' => 'high',     'pattern' => '/\b(?:cast|convert)\s*\(/i'],
        ['name' => 'sql-char-concat',       'severity' => 'high',     'pattern' => '/\bchar\s*\(\s*\d/i'],
        // Medium
        ['name' => 'sql-boolean-true',      'severity' => 'medium',   'pattern' => "/\\bor\\s+['\"\d]+\\s*=\\s*['\"\d]+/i"],
        ['name' => 'sql-boolean-and',       'severity' => 'medium',   'pattern' => "/\\band\\s+['\"\d]+\\s*=\\s*['\"\d]+/i"],
        ['name' => 'sql-order-by-num',      'severity' => 'medium',   'pattern' => '/\border\s+by\s+\d+\b/i'],
        ['name' => 'sql-tautology',         'severity' => 'medium',   'pattern' => "/'\\s*or\\s*'[^']*'\\s*=\\s*'/i"],
        // Additional coverage
        ['name' => 'sql-waitfor-delay',     'severity' => 'critical', 'pattern' => '/\bwaitfor\s+delay\b/i'],
        ['name' => 'sql-pg-sleep',          'severity' => 'critical', 'pattern' => '/\bpg_sleep\s*\(/i'],
        ['name' => 'sql-hex-values',        'severity' => 'medium',   'pattern' => '/0x[0-9a-f]{4,}/i'],
        ['name' => 'sql-group-by-having',   'severity' => 'medium',   'pattern' => '/\bhaving\s+\d+\s*=\s*\d+/i'],
        ['name' => 'sql-dbms-fingerprint',  'severity' => 'medium',   'pattern' => '/\b(?:@@version|version\s*\(\s*\)|user\s*\(\s*\)|database\s*\(\s*\))\b/i'],
        ['name' => 'sql-declare-set',       'severity' => 'high',     'pattern' => '/\bdeclare\s+@\w+\b/i'],
        ['name' => 'sql-bulk-insert',       'severity' => 'critical', 'pattern' => '/\bbulk\s+insert\b/i'],
        ['name' => 'sql-openrowset',        'severity' => 'critical', 'pattern' => '/\bopenrowset\s*\(/i'],
        // Extended coverage
        ['name' => 'sql-case-when',         'severity' => 'high',     'pattern' => '/CASE\s+WHEN\s+.*\s+THEN/i'],
        ['name' => 'sql-extractvalue',      'severity' => 'critical', 'pattern' => '/EXTRACTVALUE\s*\(/i'],
        ['name' => 'sql-updatexml',         'severity' => 'critical', 'pattern' => '/UPDATEXML\s*\(/i'],
        ['name' => 'sql-sys-tables',        'severity' => 'critical', 'pattern' => '/sys\.(user_summary|processlist|statements_with_errors)/i'],
        ['name' => 'sql-gtid',              'severity' => 'critical', 'pattern' => '/GTID_SUBSET\s*\(/i'],
        ['name' => 'sql-exp-tilde',         'severity' => 'critical', 'pattern' => '/exp\(~\(/i'],
        ['name' => 'sql-polygon',           'severity' => 'high',     'pattern' => '/(polygon|geometrycollection|linestring|multipoint)\s*\(/i'],
        ['name' => 'sql-procedure-analyse', 'severity' => 'high',     'pattern' => '/procedure\s+analyse\s*\(/i'],
        ['name' => 'sql-having',            'severity' => 'high',     'pattern' => '/\bHAVING\s+\d+\s*=\s*\d+/i'],
        ['name' => 'sql-dbms-version',      'severity' => 'critical', 'pattern' => '/@@version|@@global|@@session/i'],
    ];

    public static function scan(array $sources): ?array
    {
        foreach ($sources as $label => $values) {
            $result = self::scanValues($values, $label);
            if ($result !== null) return $result;
        }
        return null;
    }

    private static function scanValues($data, string $label): ?array
    {
        if (is_string($data)) {
            return self::matchString($data, $label);
        }
        if (is_array($data)) {
            foreach ($data as $value) {
                $r = self::scanValues($value, $label);
                if ($r !== null) return $r;
            }
        }
        return null;
    }

    private static function matchString(string $value, string $label): ?array
    {
        foreach (self::$rules as $rule) {
            if (preg_match($rule['pattern'], $value, $m)) {
                return [
                    'rule'     => $rule['name'],
                    'severity' => $rule['severity'],
                    'matched'  => substr($m[0], 0, 120),
                    'source'   => $label,
                ];
            }
        }
        return null;
    }
}

File: aspnet/src/detectors/SqlInjectionDetector.cs
using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;

namespace FireWTWall.Detectors
{
    public static class SqlInjectionDetector
    {
        private static readonly (string Name, string Severity, Regex Pattern)[] Rules =
        {
            ("sql-union-select",      "critical", new Regex(@"\bunion\s+(?:all\s+)?select\b",                         RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-drop-table",        "critical", new Regex(@";\s*drop\s+table\b",                                    RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-xp-cmdshell",       "critical", new Regex(@"\bxp_cmdshell\b",                                       RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-exec",              "critical", new Regex(@"\bexec(?:ute)?\s*\(",                                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-information-schema","critical", new Regex(@"\binformation_schema\b",                                RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-sleep",             "critical", new Regex(@"\bsleep\s*\(\s*\d",                                     RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-benchmark",         "critical", new Regex(@"\bbenchmark\s*\(",                                      RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-load-file",         "critical", new Regex(@"\bload_file\s*\(",                                      RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-into-outfile",      "critical", new Regex(@"\binto\s+(?:out|dump)file\b",                          RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-sys-tables",        "critical", new Regex(@"\bsysobjects\b|\bsyscolumns\b",                        RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-comment",           "high",     new Regex(@"(?:--|/\*|\*/|#\s*$)",                                 RegexOptions.Multiline  | RegexOptions.Compiled)),
            ("sql-stacked-query",     "high",     new Regex(@";\s*(?:select|insert|update|delete|drop|alter|create|exec)\b", RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-cast-convert",      "high",     new Regex(@"\b(?:cast|convert)\s*\(",                              RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-char-concat",       "high",     new Regex(@"\bchar\s*\(\s*\d",                                     RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-boolean-true",      "medium",   new Regex(@"\bor\s+['""\d]+\s*=\s*['""\d]+",                      RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-boolean-and",       "medium",   new Regex(@"\band\s+['""\d]+\s*=\s*['""\d]+",                     RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-order-by-num",      "medium",   new Regex(@"\border\s+by\s+\d+\b",                                 RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-tautology",         "medium",   new Regex(@"'\s*or\s*'[^']*'\s*=\s*'",                            RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-waitfor-delay",     "critical", new Regex(@"\bwaitfor\s+delay\b",                                  RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-pg-sleep",          "critical", new Regex(@"\bpg_sleep\s*\(",                                      RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-hex-values",        "medium",   new Regex(@"0x[0-9a-f]{4,}",                                       RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-group-by-having",   "medium",   new Regex(@"\bhaving\s+\d+\s*=\s*\d+",                            RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-dbms-fingerprint",  "medium",   new Regex(@"\b(?:@@version|version\s*\(\s*\)|user\s*\(\s*\)|database\s*\(\s*\))\b", RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-declare-set",       "high",     new Regex(@"\bdeclare\s+@\w+\b",                                   RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-bulk-insert",       "critical", new Regex(@"\bbulk\s+insert\b",                                    RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-openrowset",        "critical", new Regex(@"\bopenrowset\s*\(",                                     RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-case-when",         "high",     new Regex(@"CASE\s+WHEN\s+.*\s+THEN",                             RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-extractvalue",      "critical", new Regex(@"EXTRACTVALUE\s*\(",                                    RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-updatexml",         "critical", new Regex(@"UPDATEXML\s*\(",                                       RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-gtid",              "critical", new Regex(@"GTID_SUBSET\s*\(",                                     RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-exp-tilde",         "critical", new Regex(@"exp\(~\(",                                             RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-polygon",           "high",     new Regex(@"(polygon|geometrycollection|linestring|multipoint)\s*\(", RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-procedure-analyse", "high",     new Regex(@"procedure\s+analyse\s*\(",                             RegexOptions.IgnoreCase | RegexOptions.Compiled)),
            ("sql-dbms-version",      "critical", new Regex(@"@@version|@@global|@@session",                        RegexOptions.IgnoreCase | RegexOptions.Compiled)),
        };

        public static DetectorResult Scan(Dictionary<string, string> sources)
        {
            foreach (var kv in sources)
            {
                var r = MatchString(kv.Value, kv.Key);
                if (r != null) return r;
            }
            return null;
        }

        private static DetectorResult MatchString(string value, string label)
        {
            foreach (var rule in Rules)
            {
                Match m = rule.Pattern.Match(value);
                if (m.Success)
                    return new DetectorResult(rule.Name, rule.Severity,
                        m.Value.Length > 120 ? m.Value.Substring(0, 120) : m.Value, label);
            }
            return null;
        }
    }
}
